nx/lib/nx.ex

```diff
@@ -2659,16 +2659,21 @@ defmodule Nx do
   """
   @doc type: :shape
   def reshape(tensor, new_shape, opts \\ []) do
-    %T{shape: old_shape, vectorized_axes: vectorized_axes} = tensor = to_tensor(tensor)
+    %T{shape: old_shape, vectorized_axes: vectorized_axes, names: old_names} =
+      tensor = to_tensor(tensor)
+
     new_names = opts[:names] || names!(new_shape)
     new_shape = if is_tuple(new_shape), do: new_shape, else: shape(new_shape)
     new_shape = Nx.Shape.reshape(old_shape, new_shape)
 
     names = Nx.Shape.named_axes!(new_names, new_shape)
 
     cond do
-      old_shape == new_shape ->
-        %{tensor | names: names}
+      old_shape == new_shape and names == old_names ->
+        # we can't assume that a given backend does not register names for tensor
+        # shapes, so we can only return a no-op if both the shape and the names
+        # are unchanged
+        tensor
```

Comment:
@josevalim 
Another fix that smells like I'm fixing the wrong place, even though this one is more justifiable.

Basically, what's happening is that at the `expand_collapsed_axes` wasn't taking effect properly in the case that the shape was the same but I wanted to change the names of the expression.

On the other hand, it does make sense that if we were to register axes' names in a given backend, suddenly receiving a new unknown name would yield a bug, so I do think this is a structural change we need to apply.

Reply:
It is expected that backends do not register names. Names are a Nx only construct?

Reply:
I didn't know that. Though if I remove this, the name from vectorization leaks, so I think that we're not dealing with names correctly in the Expr evaluation then.

Reply:
Can you provide a small test case?

Reply:
```elixir
iex(10)> defmodule M do
...(10)>     import Nx.Defn
...(10)>     
...(10)>     defn f(%{shape: {m, n}} = x) do
...(10)>         %{shape: {k, m, n}} = devec = Nx.devectorize(x)
...(10)>         
...(10)>         {
...(10)>             Nx.reshape(devec, {k, m, n}, names: [nil, nil, nil]), 
...(10)>             Nx.reshape(devec, {k, m, n}, names: [:y, nil, nil]),
...(10)>             Nx.reshape(devec, {k, 1, m, n}, names: [:y, nil, nil, nil])
...(10)>         }
...(10)>     end
...(10)>     
...(10)>     defn g(%{shape: {m, n}} = x) do
...(10)>         %{shape: {k, m, n}} = devec = Nx.devectorize(x)
...(10)>         
...(10)>         Nx.reshape(devec, {k, m, n}, names: [nil, nil, nil])
...(10)>     end
...(10)> end
warning: redefining module M (current version defined in memory)
  iex:10

warning: variable "m" is unused (if the variable is not meant to be used, prefix it with an underscore)
  iex:13: M.f/1

warning: variable "n" is unused (if the variable is not meant to be used, prefix it with an underscore)
  iex:13: M.f/1

warning: variable "m" is unused (if the variable is not meant to be used, prefix it with an underscore)
  iex:23: M.g/1

warning: variable "n" is unused (if the variable is not meant to be used, prefix it with an underscore)
  iex:23: M.g/1

{:module, M,
 <<70, 79, 82, 49, 0, 0, 12, 28, 66, 69, 65, 77, 65, 116, 85, 56, 0, 0, 1, 39,
   0, 0, 0, 30, 8, 69, 108, 105, 120, 105, 114, 46, 77, 8, 95, 95, 105, 110,
   102, 111, 95, 95, 10, 97, 116, 116, 114, ...>>, true}
iex(11)>  M.f(Nx.iota({2, 3}, vectorized_axes: [x: 1])) 
{#Nx.Tensor<
   s64[x: 1][2][3]
   [
     [
       [0, 1, 2],
       [3, 4, 5]
     ]
   ]
 >,
 #Nx.Tensor<
   s64[x: 1][2][3]
   [
     [
       [0, 1, 2],
       [3, 4, 5]
     ]
   ]
 >,
 #Nx.Tensor<
   s64[y: 1][1][2][3]
   [
     [
       [
         [0, 1, 2],
         [3, 4, 5]
       ]
     ]
   ]
 >}
iex(12)>  M.g(Nx.iota({2, 3}, vectorized_axes: [x: 1])) 
#Nx.Tensor<
  s64[x: 1][2][3]
  [
    [
      [0, 1, 2],
      [3, 4, 5]
    ]
  ]
>
```


Here you go. I ran this on Nx main. Notice that from the output of `f`, the only case where `:names` took effect was the one in which the shape changed, and therefore the Nx.Defn.Expr backend was able to take notice of the name change.

The output of `g` shows one of the same cases, just as a sanity check that it's not related to the output being a tuple.

Another fix for this would be going into the Nx.Defn.Evaluator and changing the `apply_output` function to also apply the `output` names into the `result` tensor, but I think this is something that would have to be done by every compiler.

Reply:
After changing `apply_output` to the code below (only actual change is the addition of `Nx.rename/2`)

```elixir
defp apply_output({result, _cache}, output) do
    {result, []} =
      Composite.traverse(result, output, fn
        result, [%Nx.Tensor{vectorized_axes: vectorized_axes, names: names} | acc] ->
          corrected_result = result |> Nx.vectorize(vectorized_axes) |> Nx.rename(names)
          {corrected_result, acc}
      end)

    result
  end
```

we get the correct result:

```elixir
iex(24)> t =  M.g(Nx.iota({2, 3}, vectorized_axes: [x: 1])) 
#Nx.Tensor<
  s64[1][2][3]
  [
    [
      [0, 1, 2],
      [3, 4, 5]
    ]
  ]
>
iex(25)> t =  M.f(Nx.iota({2, 3}, vectorized_axes: [x: 1])) 
{#Nx.Tensor<
   s64[1][2][3]
   [
     [
       [0, 1, 2],
       [3, 4, 5]
     ]
   ]
 >,
 #Nx.Tensor<
   s64[y: 1][2][3]
   [
     [
       [0, 1, 2],
       [3, 4, 5]
     ]
   ]
 >,
 #Nx.Tensor<
   s64[y: 1][1][2][3]
   [
     [
       [
         [0, 1, 2],
         [3, 4, 5]
       ]
     ]
   ]
 >}
 ```
 
 In other news, EXLA does work correctly, because it already forces the buffer results into the output holders which carry the axis name info, so perhaps the way to go is to apply the fix in Defn.Evaluator only :)
 
 ```
 iex(3)> t =  M.f(Nx.iota({2, 3}, vectorized_axes: [x: 1]))                     

19:13:02.598 [info] domain=elixir.xla file=tensorflow/compiler/xla/pjrt/tfrt_cpu_pjrt_client.cc line=214  TfrtCpuClient created.
{#Nx.Tensor<
   s64[1][2][3]
   EXLA.Backend<host:0, 0.200170718.1356202002.226606>
   [
     [
       [0, 1, 2],
       [3, 4, 5]
     ]
   ]
 >,
 #Nx.Tensor<
   s64[y: 1][2][3]
   EXLA.Backend<host:0, 0.200170718.1356202002.226607>
   [
     [
       [0, 1, 2],
       [3, 4, 5]
     ]
   ]
 >,
 #Nx.Tensor<
   s64[y: 1][1][2][3]
   EXLA.Backend<host:0, 0.200170718.1356202002.226608>
   [
     [
       [
         [0, 1, 2],
         [3, 4, 5]
       ]
     ]
   ]
 >}
 ```

Reply:
Yes, your changes to apply_output are correct!

Reply:
We can also mirror EXLA and do `%{output | data: tensor.data}` (i.e. replace the data instead of vectorize+rename).